Formalize search API

jpslav opened the pull request:
Transition from the prototype search query endpoint to the "real" endpoint.

* Prototype endpoints have been moved to a `PrototypeController`
* Added Swagger documentation.  Swagger JSON is available at `/api/v0/swagger`.  
* Added `openstax_swagger` gem usage (which is an extraction of code used for `a15k/interactions-api`).  This lets us document the API with Swagger and also lets us generate PORO bindings for our API interfaces (lets us parse and export JSON by just dealing with POROs)
* Wrote the "real" search endpoint with a request spec for it.  While our main goal is to search one book at a time, the search endpoint takes an array of books to make it possible to search across books should we want to in the future.
* Added "search strategies" to complement "index strategies".  Whereas index strategies control how a book is indexed into Elasticsearch, search strategies control how we query that content out of Elasticsearch.  Implemented the "S1" search strategy to match the search requirements from UX.  See the [simple_query_string](https://www.elastic.co/guide/en/elasticsearch/reference/6.3/query-dsl-simple-query-string-query.html) reference page for details on the options we are using.
* Currently, the index strategy and search strategy are provided as inputs to the search endpoint.  At some point in the future we may not do that, but for now it gives us flexibility as we experiment with search.
* Added a rake task to generate the files with the Swagger JSON (see the README)

(rest of todos moved to another separate issue)

jpslav committed:
move prototype code into a prototype controller

jpslav committed:
speed up boot time

jpslav committed:
swagger models and start of controller docs

jpslav committed:
add search strategies and use in search controller

TomWoodward commented on app/controllers/api/v0/search_controller.rb:
```diff
@@ -1,47 +1,64 @@
+        key :type, :string
+      end
+      parameter do
+        key :name, :index_strategy
```
didn't we discuss this being fixed per api version and the client not needing to care about it?

jpslav replied:
Let's maybe park this.  I do recall some conversation related to this, but there are reasons to not have a different API version for every index strategy / search strategy combination.  As it is currently...

* The API version controls the format of the requests and the top-level format of the response
* The index strategy controls what is indexed and how it is indexed (e.g. page elements, indexing options, etc)
* The search strategy controls how we search the index (fuzzy vs not fuzzy, boosting fields, full matches vs OR matches, etc)

The API is returning raw ES results plus some top-level metadata.  The top-level metadata is fixed for each API version.  The raw ES results depend on the combination of strategies.

One API version, defining the request format and top-level response format, could work with several combinations of strategies.  

There likely won't be a zillion strategies, and the Rex FE would change slowly over time to use different ones (e.g. the `i1` indexing strategy could be used for many months).  

The `search_strategy` parameter would give us a way to A/B test different search queries (with the A/B machinery being external to search itself).

TomWoodward replied:
i was thinking that indexing/search strategy changes that did not influence the structure of the responses could be done within one api version, and any other change would create an api version. exposing 3 versions of things for the client to manage seems like a lot, and i was actually thinking it would be nicer to handle a/b testing of search stuff within search, rather than forcing the client to deal with all that

jpslav replied:
After a discussion with Tom, we decided it is ok for the time being to leave the query parameters that choose index strategy and search strategy.

TomWoodward commented on app/controllers/api/v0/search_controller.rb:
```diff
@@ -1,47 +1,64 @@
+        key :type, :string
+      end
+      parameter do
+        key :name, :search_strategy
```
same for this?

TomWoodward commented on app/controllers/api/v0/search_controller.rb:
```diff
@@ -1,47 +1,64 @@
-
-    response = OpenSearch::ElasticsearchClient.instance.search body: query.to_json
-    render json: response
+  swagger_path '/search' do
```
so this path is going to end up looking something like `openstax.org/search/api/v0/search` right?

jpslav replied:
Yes... `openstax.org/open-search/api/v0/search`.  The `/api/v0` part is defined as an API base somewhere else... but this is a good point that we don't have the `open-search` prefix there.  We could add it and just make all Swagger use assume access via CloudFront (and that prefix) or we could have two Swagger JSONs, one with and one without the prefix.

TomWoodward replied:
we can just use the prefix all the time right? no reason to support queries without the prefix like accounts does

jpslav replied:
I updated to use the `/open-search` prefix in the swagger json

TomWoodward reviewed (commented):
i in general don't love committing generated code, but i'll defer to you ruby guys

jpslav commented on app/controllers/api/v0/search_controller.rb:
```diff
@@ -1,47 +1,64 @@
+        key :type, :string
+      end
+      parameter do
+        key :name, :book
```
You mean searching across multiple books?  We don't have that use case envisioned do we?

We could make the parameter be an array and freak out if there's more than one entry for the time being.

TomWoodward replied:
i don't think UI has any plans to implement, but if it was easy to do on the back we could just do it, if not, whatever

jpslav replied:
I updated this to take a list of books, and it *should* work, tho it isn't tested :-o

jpslav committed:
search query specs and real search strategy matching requirements

jpslav committed:
update aws-ruby and cnx-ruby to get rid of an error and a warning

jpslav committed:
fix scoping bug seen in prod

jpslav committed:
don't modify dynamodb table name

jpslav committed:
new version of swagger-rails

jpslav committed:
increase worker capacity to handle new jobs

jpslav committed:
fix bad arg bug

jpslav committed:
don't search hidden content for now

jpslav committed:
lifecycle hook mechanics for enqueue index jobs rake task

jpslav committed:
describe the ES results in swagger

jpslav committed:
more required fields in search response

jpslav committed:
templatized worker asg name secret

steveburkett committed:
Merge branch 'master' into api-1

steveburkett committed:
Fix specs broken due to merge

steveburkett committed:
Use element ID in search

steveburkett committed:
Correct the swagger ruby bindings for adding the element ID

jpslav committed:
fix install_secrets to handle multi-level secrets namespaces

steveburkett committed:
Changed object interrogation to raise exception

It'll be rare that an element doesnt have an ID, so raise an exception
if it's nil. Rescue the exception and sent to sentry

jpslav committed:
Merge pull request #28 from openstax/sb/add_element_id_to_search

Add element id to search

steveburkett committed:
Fix failing spec for ElementIdMissing

jpslav committed:
Merge pull request #30 from openstax/sb/fix_failing_spec

Fix failing spec for ElementIdMissing

jpslav committed:
Merge pull request #29 from openstax/fix-install-secrets-for-split-namespace

fix install_secrets to handle multi-level secrets namespaces

jpslav committed:
make workers keep looking for termination

jpslav committed:
more logging in worker rake task

jpslav committed:
send exceptions to sentry right away, don't wait for lighter reporting via done job

steveburkett commented on config/initializers/aws.rb:
```diff
@@ -12,8 +12,20 @@ def aws_creds
-  {
+def use_aws?
+  !Rails.env.development? ||
+  ENV['USE_AWS'] ||
```
not sure if USE_AWS is needed....

steveburkett commented on config/initializers/openstax_swagger.rb:
```diff
@@ -0,0 +1,10 @@
+    )
+  }
+end
+
```
extra lines

steveburkett commented on spec/requests/api/v0/search_requests_spec.rb:
```diff
@@ -0,0 +1,58 @@
+        element_type: "paragraph",
+        page_position: 3
+      )
+      expect(json[:hits][:hits][0][:highlight][:visible_content][0]).to start_with "<em>Recall</em>"
```
uggggg

steveburkett committed:
Improve rex release code to allow for N level nested rex release

jpslav committed:
Merge pull request #32 from openstax/sb/flexible_rex_release_dir_structure

Improve rex release code to allow for N level nested rex release

jpslav commented on app/controllers/api/v0/prototype_controller.rb:
```diff
@@ -0,0 +1,47 @@
+class Api::V0::PrototypeController < Api::V0::BaseController
+
+  def search
+    # TODO generalize
```
This class will be deleted in a couple of weeks

jpslav commented on app/controllers/api/v0/prototype_controller.rb:
```diff
@@ -0,0 +1,47 @@
+    render json: response
+  end
+
+  def temp_build_index
```
This class will be deleted in a couple of weeks

jpslav commented on Gemfile:
```diff
@@ -45,7 +45,7 @@ gem 'dotenv-rails'
 gem "openstax_path_prefixer", github: "openstax/path_prefixer", ref: "b6d8f45d8"
 
-gem "openstax_aws", github: "openstax/aws-ruby", ref: '1554be9d36df9f8129'
+gem "openstax_aws", github: "openstax/aws-ruby", ref: '1554be9d36'
```
I think 7 digits or more is good.  More digits makes (extremely unlikely) collisions less probable.

steveburkett committed:
Add rescue in monitor jobs & fix specs

jpslav committed:
namespace temp aws envs

steveburkett committed:
new recordings

jpslav committed:
Merge pull request #33 from openstax/sb/fix-travis-rescue-in-monitor

Fix travis errors & add rescue in monitor

jpslav committed:
Merge pull request #31 from openstax/no-zombie-workers

make workers keep looking for termination
